### PR TITLE
Ahmaalba/least privilege design

### PR DESCRIPTION
**Pull request by abassyiouni**
**Description**

- Separated Adapter Role and Engine Role
- Removed S3ReadOnly, and EFSFull managed policies, and instead replaced them with more restrictive policies.

**Testing**

- All code tested `make` as well activating, and deploying contexts.
- Successfully ran `hello`, `seq-format-validation`, `bam-to-unmapped-bams`, `cram-to-bam cromwell` workflows
- Successfully ran `hello`,`rnaseq` nextflow workflows

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts [hunk 1]
```diff
@@ -13,23 +13,32 @@ import { EngineOptions, ServiceContainer } from "../../types";
...
 import { EngineOutputs, NestedEngineStack } from "./nested-engine-stack";
+import { CromwellEngineRole } from "../../roles/cromwell-engine-role";
 
 export interface CromwellEngineStackProps extends EngineOptions, NestedStackProps {}
 
 export class CromwellEngineStack extends NestedEngineStack {
   public readonly engine: SecureService;
   public readonly adapter: SecureService;
-  public readonly taskRole: IRole;
+  public readonly adapterRoll: IRole;
```
Comment (IllyaYalovyy) [inline packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts:23]: typo: "adapterRole"

File: packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts [hunk 2]
```diff
@@ -13,23 +13,32 @@ import { EngineOptions, ServiceContainer } from "../../types";
...
     const engineContainer = params.getEngineContainer(props.jobQueue.jobQueueArn);
-    this.taskRole = new Role(this, "TaskRole", { assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"), ...props.policyOptions });
+    const artifactBucket = Bucket.fromBucketName(this, "ArtifactBucket", params.artifactBucketName);
+    const outputBucket = Bucket.fromBucketName(this, "OutputBucket", params.outputBucketName);
+    this.engineRole = new CromwellEngineRole(this, "CromwellEngineRole", {
+      readOnlyBucketArns: (params.readBucketArns ?? []).concat(artifactBucket.bucketArn),
+      readWriteBucketArns: (params.readWriteBucketArns ?? []).concat(outputBucket.bucketArn),
+      policies: props.policyOptions,
+    });
+    this.adapterRoll = new Role(this, "TaskRole", { assumedBy: new ServicePrincipal("ecs-tasks.amazonaws.com"), ...props.policyOptions });
```
Comment (IllyaYalovyy) [inline packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts:41]: ditto

File: packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts [hunk 3]
```diff
@@ -38,20 +47,19 @@ export class CromwellEngineStack extends NestedEngineStack {
       name: engineContainer.serviceName,
       cloudMapNamespace: namespace,
     };
+
     // TODO: Move log group creation into service construct and make it a property
-    this.engineLogGroup = new LogGroup(this, "EngineLogGroup");
     this.engine = this.getEngineServiceDefinition(props.vpc, engineContainer, cloudMapOptions, this.engineLogGroup);
     this.adapterLogGroup = new LogGroup(this, "AdapterLogGroup");
-    this.adapter = renderServiceWithContainer(this, "Adapter", params.getAdapterContainer(), props.vpc, this.taskRole, this.adapterLogGroup);
+    this.adapter = renderServiceWithContainer(this, "Adapter", params.getAdapterContainer(), props.vpc, this.adapterRoll, this.adapterLogGroup);
```
Comment (IllyaYalovyy) [inline packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts:54]: ditto

File: packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts [hunk 4]
```diff
@@ -38,20 +47,19 @@ export class CromwellEngineStack extends NestedEngineStack {
...
 
     this.apiProxy = new ApiProxy(this, {
       apiName: `${params.projectName}${params.contextName}${engineContainer.serviceName}ApiProxy`,
       loadBalancer: this.adapter.loadBalancer,
       allowedAccountIds: [this.account],
     });
 
-    const outputBucket = Bucket.fromBucketName(this, "OutputBucket", params.outputBucketName);
-    outputBucket.grantReadWrite(this.taskRole);
+    outputBucket.grantReadWrite(this.adapterRoll);
```
Comment (IllyaYalovyy) [inline packages/cdk/lib/stacks/nested/cromwell-engine-stack.ts:62]: ditto

**Review by IllyaYalovyy — CHANGES_REQUESTED** [review]
Thank you for this great contribution. Just some trivial typos to take care of.

**Review by IllyaYalovyy — APPROVED** [review]

**Review by markjschreiber — APPROVED** [review]

**Review by AndreyDovydenko — APPROVED** [review]

**Review by AndreyDovydenko — APPROVED** [review]

**Pull request merged**